pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,287 @@
+	ctx           = context.Background()
+	result        = reconcile.Result{}
+	resultRequeue = reconcile.Result{Requeue: true}
+)
+
+// GKEClusterController is responsible for adding the GKECluster
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type GKEClusterController struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *GKEClusterController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1beta1.GKEClusterGroupVersionKind),
+		resource.WithExternalConnecter(&clusterConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}),
+		resource.WithManagedInitializers())
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1beta1.GKEClusterKindAPIVersion, v1beta1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.GKECluster{}).
+		Complete(r)
+}
+
+type clusterConnector struct {
+	kube         client.Client
+	newServiceFn func(ctx context.Context, opts ...option.ClientOption) (*container.Service, error)
+}
+
+func (c *clusterConnector) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return nil, errors.New(errNotCluster)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	gke, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &clusterExternal{cluster: *gke, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type clusterExternal struct {
+	kube      client.Client
+	cluster   container.Service
+	projectID string
+}
+
+func (e *clusterExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotCluster)
+	}
+
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(meta.GetExternalName(mg)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetCluster)
+	}
+
+	cr.Status.AtProvider = gke.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	gke.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !reflect.DeepEqual(currentSpec, &cr.Spec.ForProvider) {
```

Comment:
Should we use cmp.Equal here as we do in the client comparison functions?

Reply:
That sounds fine to me. I had noticed @muvaf used `DeepEqual` for redis and cloudsql, @muvaf was there any specific reason for that over `cmp.Equal`?

Reply:
Here is why cmp package exists:
```
// This package is intended to be a more powerful and safer alternative to
// reflect.DeepEqual for comparing whether two values are semantically equal.
//
// The primary features of cmp are:
//
// • When the default behavior of equality does not suit the needs of the test,
// custom equality functions can override the equality operation.
// For example, an equality function may report floats as equal so long as they
// are within some tolerance of each other.
//
// • Types that have an Equal method may use that method to determine equality.
// This allows package authors to determine the equality operation for the types
// that they define.
//
// • If no custom equality functions are used and no Equal method is defined,
// equality is determined by recursively comparing the primitive kinds on both
// values, much like reflect.DeepEqual. Unlike reflect.DeepEqual, unexported
// fields are not compared by default; they result in panics unless suppressed
// by using an Ignore option (see cmpopts.IgnoreUnexported) or explicitly compared
// using the AllowUnexported option.
```

I remember thinking that I don't need any of those features, at least at that time, so I went with using the standard library for non-test code. But it should be fine to use it if you'd like. It's imported for the tests anyway, though I'm not 100% sure that if tests import a package it ends up compiled in the final binary.

Reply:
I like that `cmp` ignores unexported fields, and allows custom equality override logic. Given `cmp` and `DeepEqual` do the same thing in most case, but `cmp` _can_ do more, my preference is to just default to using `cmp` for comparison.